src/triton_fil/triton_tensor.cuh

```diff
@@ -56,38 +60,95 @@ class TritonTensor {
-            auto cur_head = ptr_d;
-            for (auto& buffer_ : buffers) {
-              try {
-                raft::copy(
-                    cur_head, reinterpret_cast<const byte*>(buffer_.data),
-                    buffer_.size_bytes, stream_);
+            byte* ptr_d;
+            if (target_memory == TRITONSERVER_MEMORY_GPU) {
+              ptr_d = allocate_device_memory<byte>(size_bytes_);
+              auto cur_head = ptr_d;
+              for (auto& buffer_ : buffers) {
+                try {
+                  raft::copy(
+                      cur_head, reinterpret_cast<const byte*>(buffer_.data),
+                      buffer_.size_bytes, stream_);
+                }
+                catch (const raft::cuda_error& err) {
+                  throw TritonException(
+                      TRITONSERVER_errorcode_enum::TRITONSERVER_ERROR_INTERNAL,
+                      err.what());
+                }
+                cur_head += buffer_.size_bytes;
               }
-              catch (const raft::cuda_error& err) {
-                throw TritonException(
-                    TRITONSERVER_errorcode_enum::TRITONSERVER_ERROR_INTERNAL,
-                    err.what());
+            } else if (target_memory == TRITONSERVER_MEMORY_CPU) {
+
+              LOG_MESSAGE(TRITONSERVER_LOG_INFO, "TritonTensor::TritonTensor()");
+
+              ptr_d = static_cast<byte*>(std::malloc(size_bytes_ * sizeof(byte)));
+
+              LOG_MESSAGE(TRITONSERVER_LOG_INFO, (std::string("size_bytes_ = ") + std::to_string(size_bytes_)).c_str());
+
+              auto cur_head = ptr_d;
+              for (auto& buffer_ : buffers) {
+
+                LOG_MESSAGE(TRITONSERVER_LOG_INFO, TRITONSERVER_MemoryTypeString(buffer_.memory_type));
+
+                cudaPointerAttributes s_att;
+                cudaError_t s_err = cudaPointerGetAttributes(&s_att, buffer_.data);
+                if (s_err != cudaSuccess) {
+                  std::string msg{"Call to cudaPointerGetAttributes() failed"};
+                  msg += cudaGetErrorString(s_err);
+                  throw TritonException(
+                      TRITONSERVER_errorcode_enum::TRITONSERVER_ERROR_INTERNAL, msg);
+                }
+
+                if (s_att.type == cudaMemoryTypeDevice) {
+                  LOG_MESSAGE(TRITONSERVER_LOG_INFO, "buffer is on device");
+                  LOG_MESSAGE(TRITONSERVER_LOG_INFO, (std::string("buffer_.size_bytes = ") + std::to_string(buffer_.size_bytes)).c_str());
+                  CUdeviceptr pbase;
+                  std::size_t psize;
+                  cuMemGetAddressRange(&pbase, &psize, (CUdeviceptr)(uintptr_t)(buffer_.data));
+                  LOG_MESSAGE(TRITONSERVER_LOG_INFO, (std::string("buffer_.data() has size ") + std::to_string(psize)).c_str());
+                  raft::copy(
+                      cur_head, reinterpret_cast<const byte*>(buffer_.data),
+                      buffer_.size_bytes, stream_);
```

Comment:
There is a bug in how `buffer_.size_bytes` is set. The log prints like this:
```
I0603 20:05:34.847834 1 triton_tensor.cuh:113] GPU
I0603 20:05:34.847843 1 triton_tensor.cuh:125] buffer is on device
I0603 20:05:34.847852 1 triton_tensor.cuh:126] buffer_.size_bytes = 512000
I0603 20:05:34.847862 1 triton_tensor.cuh:130] buffer_.data() has size 256000
```

Reply:
`cudaMemcpyAsync` returns `cudaErrorInvalidValue` if the user attempts to copy more bytes than there are available from the device memory. See https://nanxiao.me/en/the-invalid-argument-error-of-cudamemcpyasync/